packages/@lwc/engine-core/src/framework/freeze-template.ts

```diff
@@ -81,48 +83,85 @@ function warnOnArrayMutation(stylesheets: TemplateStylesheetFactories) {
+
             set(value) {
-                logError(
-                    `Dynamically setting the "stylesheetTokens" property on a template function ` +
-                        `is deprecated and may be removed in a future version of LWC.`
-                );
-                // Avoid logging twice (for both stylesheetToken and stylesheetTokens)
-                mutationWarningsSilenced = true;
-                originalDescriptor!.set!.call(this, value);
-                mutationWarningsSilenced = false;
+                // If the value is null or some other exotic object, you would be broken anyway in the past
+                // because the engine would try to access hostAttribute/shadowAttribute, which would throw an error.
+                // However it may be undefined in newer versions of LWC, so we need to guard against that case.
+                this.stylesheetToken = isUndefined(value)
+                    ? undefined
+                    : (value as any).shadowAttribute;
             },
         });
+
+        // When ENABLE_FROZEN_TEMPLATE is false, warn in dev mode whenever someone is mutating the template
+        if (process.env.NODE_ENV !== 'production') {
+            if (!isUndefined(tmpl.stylesheets)) {
+                warnOnArrayMutation(tmpl.stylesheets);
+            }
+            for (const prop of TEMPLATE_PROPS) {
+                let value = tmpl[prop];
+                defineProperty(tmpl, prop, {
+                    enumerable: true,
+                    configurable: true,
+                    get() {
+                        return value;
+                    },
+                    set(newValue) {
+                        if (!mutationWarningsSilenced) {
+                            logError(
+                                `Dynamically setting the "${prop}" property on a template function ` +
+                                    `is deprecated and may be removed in a future version of LWC.`
+                            );
+                        }
+                        value = newValue;
+                    },
+                });
+            }
+
+            const originalDescriptor = getOwnPropertyDescriptor(tmpl, 'stylesheetTokens');
+            defineProperty(tmpl, 'stylesheetTokens', {
+                enumerable: true,
+                configurable: true,
+                get: originalDescriptor!.get,
+                set(value) {
+                    logError(
+                        `Dynamically setting the "stylesheetTokens" property on a template function ` +
+                            `is deprecated and may be removed in a future version of LWC.`
+                    );
+                    // Avoid logging twice (for both stylesheetToken and stylesheetTokens)
+                    mutationWarningsSilenced = true;
+                    originalDescriptor!.set!.call(this, value);
+                    mutationWarningsSilenced = false;
+                },
+            });
+        }
```

Comment:
This diff is really confusing. I didn't actually change this code; I just moved code from `secure-template.ts` to this file.